libs/langchain/langchain/document_loaders/recursive_url_loader.py

```diff
@@ -14,19 +33,114 @@ def __init__(
         self,
         url: str,
         exclude_dirs: Optional[str] = None,
+        use_async: bool = False,
+        extractor: Callable[[str], str] = lambda x: x,
+        max_depth: int = 2,
+        timeout: int = 10,
+        prevent_outside: bool = True,
     ) -> None:
         """Initialize with URL to crawl and any subdirectories to exclude.
-
         Args:
             url: The URL to crawl.
             exclude_dirs: A list of subdirectories to exclude.
+            use_async: Whether to use asynchronous loading,
+            if use_async is true, this function will not be lazy,
+            but it will still work in the expected way, just not lazy.
+            extractor: A function to extract the text from the html,
+            when extract function returns empty string, the document will be ignored.
+            max_depth: The max depth of the recursive loading.
+            timeout: The timeout for the requests, in the unit of seconds.
         """
 
         self.url = url
         self.exclude_dirs = exclude_dirs
+        self.use_async = use_async
+        self.extractor = extractor
+        self.max_depth = max_depth
+        self.timeout = timeout
+        self.prevent_outside = prevent_outside
+
+    def _get_sub_links(self, raw_html: str, base_url: str) -> List[str]:
+        """This function extracts all the links from the raw html,
+        and convert them into absolute paths.
+
+        Args:
+            raw_html (str): original html
+            base_url (str): the base url of the html
+
+        Returns:
+            List[str]: sub links
+        """
+        # Get all links that are relative to the root of the website
+        all_links = re.findall(r"href=[\"\'](.*?)[\"\']", raw_html)
+        absolute_paths = []
+        # Process the links
+        for link in all_links:
+            # Some patterns are on blacklist
+            # like javascript: or mailto:
+            if link.startswith("javascript:") or link.startswith("mailto:"):
+                continue
+            # Or #, which is a link to the same page
+            if link.startswith("#"):
+                continue
+            # Some may be css, js, images files
+            invalid_ext = [
+                ".css",
+                ".js",
+                ".ico",
+                ".png",
+                ".jpg",
+                ".jpeg",
+                ".gif",
+                ".svg",
+            ]
+            if any([link.endswith(ext) for ext in invalid_ext]):
+                continue
+            # Some may be absolute links like https://to/path
+            if link.startswith("http"):
+                if (not self.prevent_outside) or (
+                    self.prevent_outside and link.startswith(base_url)
+                ):
+                    absolute_paths.append(link)
+            else:
```

Comment:
This always adds relative paths then?

Reply:
Yes, relative paths are certainly sub-domains of the target URL, so it's our target.